src/hierarchies/lapnig_hierarchy.cc

```diff
@@ -0,0 +1,203 @@
+  auto &rng = bayesmix::Rng::Instance().get();
+  LapNIG::State out{};
+  out.scale = stan::math::inv_gamma_rng(params.shape, 1. / params.scale, rng);
+  out.mean = stan::math::normal_rng(params.mean, sqrt(params.var), rng);
+  return out;
+}
+
+void LapNIGHierarchy::update_summary_statistics(
+    const Eigen::RowVectorXd &datum, bool add) {
+  if (add) {
+    cluster_data_values.push_back(datum);
+  } else {
+    auto it = std::find(cluster_data_values.begin(), cluster_data_values.end(),
+                        datum);
+    cluster_data_values.erase(it);
+  }
+}
+
+void LapNIGHierarchy::sample_full_cond(bool update_params) {
+  if (this->card == 0) {
+    // No posterior update possible
+    this->sample_prior();
+  } else {
+    // Number of iterations to compute the acceptance rate
+    ++iter_;
+
+    // Random generator
+    auto &rng = bayesmix::Rng::Instance().get();
+
+    // Candidate mean and candidate log_scale
+    Eigen::VectorXd curr_unc_params(2);
+    curr_unc_params << state.mean, std::log(state.scale);
+
+    Eigen::VectorXd prop_unc_params = propose_rwmh(curr_unc_params);
+
+    double log_target_prop = eval_prior_lpdf_unconstrained(prop_unc_params) +
+                             eval_like_lpdf_unconstrained(prop_unc_params);
+
+    double log_target_curr = eval_prior_lpdf_unconstrained(curr_unc_params) +
+                             eval_like_lpdf_unconstrained(curr_unc_params);
+
+    double log_a_rate = log_target_prop - log_target_curr;
+
+    if (std::log(stan::math::uniform_rng(0, 1, rng)) < log_a_rate) {
+      ++accepted_;
+      state.mean = prop_unc_params(0);
+      state.scale = std::exp(prop_unc_params(1));
+    }
+  }
+}
+
+Eigen::VectorXd LapNIGHierarchy::propose_rwmh(
+    const Eigen::VectorXd &curr_vals) {
+  auto &rng = bayesmix::Rng::Instance().get();
+  double candidate_mean =
+      curr_vals(0) + stan::math::normal_rng(0, sqrt(hypers->mh_mean_var), rng);
+  double candidate_log_scale =
+      curr_vals(1) +
+      stan::math::normal_rng(0, sqrt(hypers->mh_log_scale_var), rng);
+  Eigen::VectorXd proposal(2);
+  proposal << candidate_mean, candidate_log_scale;
+  return proposal;
+}
+
+double LapNIGHierarchy::eval_prior_lpdf_unconstrained(
+    Eigen::VectorXd unconstrained_parameters) {
+  double mu = unconstrained_parameters(0);
+  double log_scale = unconstrained_parameters(1);
+  double scale = std::exp(log_scale);
+  return stan::math::normal_lpdf(mu, hypers->mean, std::sqrt(hypers->var)) +
+         stan::math::inv_gamma_lpdf(scale, hypers->shape, hypers->scale) +
+         log_scale;
+}
+
+double LapNIGHierarchy::eval_like_lpdf_unconstrained(
+    Eigen::VectorXd unconstrained_parameters) {
+  double mean = unconstrained_parameters(0);
+  double log_scale = unconstrained_parameters(1);
+  double scale = std::exp(log_scale);
+  double diff_sum = 0;  // Sum of absolute values of data - candidate_mean
+  for (auto &elem : cluster_data_values) {
+    diff_sum += std::abs(elem(0, 0) - mean);
+  }
```

Comment:
This doesn't strike me as being very efficient... Is there any way to track the current `diff_sum` value without having to store all the data and perform this loop every time? I'm asking since you probably have already thought about this

Reply:
We compute the sum of absolute differences every time when we evaluate the log-likelihood. When we do it for the proposal mean it is unavoidable, since we have different value for the mean every time. For the current mean, we found a workaround to be more efficient. We introduced 2 new members: sum_abs_diff_curr & sum_abs_diff_prop. We update sum_abs_diff_curr every time we add/remove data and when the proposal is accepted.
What do you think?

Reply:
That's very clever :)